process/best_practices_badge.md

```diff
@@ -0,0 +1,57 @@
+
+[https://sonarcloud.io/organizations/academysoftwarefoundation/projects](https://www.google.com/url?q=https://sonarcloud.io/organizations/academysoftwarefoundation/projects&sa=D&source=docs&ust=1675375752427760&usg=AOvVaw3aJ1DsMv9yF2Xf-Uhv8EKo)
+
+I'm also seeing some work towards using clang-tidy in Imath: [https://cs.github.com/AcademySoftwareFoundation/Imath/blob/3ad5d4da02795dbf37b44c5bbd2f919a7c01a19a/config/ImathSetup.cmake#L118](https://www.google.com/url?q=https://cs.github.com/AcademySoftwareFoundation/Imath/blob/3ad5d4da02795dbf37b44c5bbd2f919a7c01a19a/config/ImathSetup.cmake%23L118&sa=D&source=docs&ust=1675375752432129&usg=AOvVaw1HThXkwYas6U7fU6_IKetY)
+
+and in OSL:
+
+[https://cs.github.com/AcademySoftwareFoundation/Imath/blob/3ad5d4da02795dbf37b44c5bbd2f919a7c01a19a/config/ImathSetup.cmake#L118](https://www.google.com/url?q=https://cs.github.com/AcademySoftwareFoundation/Imath/blob/3ad5d4da02795dbf37b44c5bbd2f919a7c01a19a/config/ImathSetup.cmake%23L118&sa=D&source=docs&ust=1675375752432249&usg=AOvVaw10Qy3WPYpoaDl6XTJfp3F9)
+
+## "The project MUST have a documented process for responding to vulnerability reports." - does GitHub have anything to help with this?
+
+
+
+## "The project MUST list external dependencies in a computer-processable way." - is there an automated way to build this?
+
+LFX Security automatically will pull the list of dependencies for a project provided they use the standard mechanisms leveraged by Synk.
+
+## "The project MUST provide an assurance case that justifies why its security requirements are met. The assurance case MUST include: a description of the threat model, clear identification of trust boundaries, an argument that secure design principles have been applied, and an argument that common implementation security weaknesses have been countered."
+
+
+
+## The project MUST have a [reproducible build](https://reproducible-builds.org/). If no building occurs (e.g., scripting languages where the source code is used directly instead of being compiled), select "not applicable"
+
+The easiest solution is not to provide a binary build of the project, which is generally recommended. 
+
+## The project MUST have performed a security review within the last 5 years. This review MUST consider the security requirements and security boundary. 
+
+
+
+## "The project MUST have FLOSS automated test suite(s) that provide at least 90% statement coverage if there is at least one FLOSS tool that can measure this criterion in the selected language. "and "The project MUST have FLOSS automated test suite(s) that provide at least 80% branch coverage if there is at least one FLOSS tool that can measure this criterion in the selected language."
+
```

Comment:
Would be good to have a standard way for projects to represent this information on their README.md, and possibly feed the results back to the LFX dashboard.

Reply:
If everyone uses SonarCloud, I believe they have badges that can be used here.

Reply:
90% is extremely hard to achieve. I've spent months (as a background task) struggling to get OSL and OIIO from 60% or so range they started in to above 80%. Going from 60-70 was easy, the higher over 70 I got, I had to fight hard for every additional 0.1%. I'm glad I did it, I am happy with the additional tests I've added and the bugs it has revealed/fixed along the way. But going from 80 to 90 may be extremely difficult if not impossible in practice, and almost certainly is hitting a point of diminish returns, versus spending equivalent effort fixing *reported* bugs and adding features.

Reply:
So I recall this coming up, and one of the challenges of 90% was getting into oddball code paths that might be complier option dependent or really niche use cases - would you say that is the challenge here?

Reply:
I think there are several issues. Some individual code paths are hard to test for obvious reasons:

* Code that handles errors that are very hard to replicate on purpose, or hard to replicate on the runners. How on the GHA runner do we trigger the error that only happens when the disk volume is full? Or a corner case that can only happen when reading from a 2TB file?

* Code that only comes into play when there is live user interaction. (I cheat by excluding the interactive apps from the coverage analysis, but that artificially boosts my test percentage, because I'm excluding some things from the denominator.)

* Sometimes multiple code paths are mutually exclusive in any build or any run, so to actually have coverage of all the options, I'd need to have multiple separate static analysis jobs and/or full testsuite runs in order to truly have full coverage.

But the most difficult phenomenon is simply the diminishing returns of each unit of work when writing more tests. When you're at 50% coverage, you look at the coverage reports and say "oh, this block of 1000 lines that is untested only is needed for multi-part exr files, I must not have any of those in my testsuite", so you add a new test for that, and suddenly your coverage jumps by 3% just from that alone.  But somewhat later, you're at 75% coverage, and you realize that there are no more big blocks of code that are untested, you just have thousands of completely independent 5-line sections that are untested, unrelated to each other, so every new test case you might work hard to add will only improve your coverage by 0.01%. Going from 50% to 60% was less work than going from 80% to 80.5%.

Reply:
First off - I agree 90% test coverage is a bear. I did TDD in my former life, mostly against legacy codebases, and the amount of scaffolding to get some tests to run can be more complicated to manage than the code itself. I could add all the TDD fanboy-isms here - but I'm sure you've heard it all before ;-)

If you look at the aim of the gold level badge requirements, it's to position a project in the best way to be sustainable and maintained in a way that downstream users can have trust in the project. I look at the test coverage requirement through that lens; it's all about bringing a higher level of rigor to the project that, in turn, signals to downstream users the project's maturity. 

For each project going through the badging process, all have had to make some changes to what they had to hit the requirements, and for many, this was a lot of work. So I guess it's an expectation that 90% code coverage is hard to get to because the gold level badge is designed to be hard to achieve ;-). But in the end, having that level of coverage does make the project better, just as all the requirements do.

Reply:
I added some context on testing integration to third-party hardware/applications, as well the case of build/compile options changing coverage and how to best measure that.